### PR TITLE
dawn: add decentralized wifi controller

### PR DESCRIPTION
**Pull request by PolynomialDivision**
Dawn is a decentralized WiFi controller.
Just install dawn and the APs will find each other via umdns.
They periodically exchange information about connected clients, wireless
statistics and other needed information. With that, the daemon load
balances clients between different APs through association control.

Further, the daemon exposes through ubus the hearing map and the
complete wireless network overview. The hearing map is the list of all
probe requests seen from a client from all APs that are running the
controller.

Hearing map:

    ubus call dawn get_hearing_map

Network overview:

    ubus call dawn get_network

Dawn has a graphical user interface called luci-app-dawn. With that, the
load balancing and other useful settings can be controlled.

Next steps:
- [ ] Neighborreports
- [ ] BSS Trasnsition

Adding the controller to official openwrt package tree was wished in 
https://github.com/berlin-open-wireless-lab/DAWN/issues/53

If I should maintain the controller here, I would like commit priviliges. ;)

### CODE REVIEW

**Comment by PolynomialDivision** [comment]
I will add luci-app-dawn after the controller is accepted.

**Comment by neheb** [comment]
Clean. The only suggestion I have is adding PKG_BUILD_PARALLEL for faster compilation.

As a sidenote, I've been meaning to convert cmake.mk to use Ninja but have not gotten far..

File: net/dawn/Makefile [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+CMAKE_INSTALL:=1
+
+include $(INCLUDE_DIR)/package.mk
+include $(INCLUDE_DIR)/cmake.mk
+
+define Package/dawn
+	SECTION:=net
+	CATEGORY:=Network
+	TITLE:=Decentralized wifi controller.
+	MAINTAINER:=Nick Hainke <vincent@systemli.org>
```
Comment (BKPepe) [inline net/dawn/Makefile]: This is already on 16th line.

File: net/dawn/Makefile [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+PKG_BUILD_PARALLEL:=1
+CMAKE_INSTALL:=1
+
+include $(INCLUDE_DIR)/package.mk
+include $(INCLUDE_DIR)/cmake.mk
+
+define Package/dawn
+	SECTION:=net
+	CATEGORY:=Network
+	TITLE:=Decentralized wifi controller.
```
Comment (BKPepe) [inline net/dawn/Makefile]: Remove dot.

File: net/dawn/Makefile [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+PKG_LICENSE:=GPL-2.0
+PKG_LICENSE_FILES:=COPYING
+
+PKG_BUILD_PARALLEL:=1
+CMAKE_INSTALL:=1
+
+include $(INCLUDE_DIR)/package.mk
+include $(INCLUDE_DIR)/cmake.mk
+
+define Package/dawn
```
Comment (BKPepe) [inline net/dawn/Makefile]: This define should have two spaces as indentation.

File: net/dawn/Makefile [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+PKG_VERSION:=2020-03-23
+PKG_RELEASE:=1
+
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_URL:=https://github.com/berlin-open-wireless-lab/DAWN.git
+PKG_SOURCE_VERSION:=42be2efdaaef0f8e7d204f558610a494577b14ec
+PKG_MIRROR_HASH:=aa3c5970cd853bd4db4ffa3868f7c1bfd0ad47f02a97a31fc093e4d24e75512d
+
+PKG_MAINTAINER:=Nick Hainke <vincent@systemli.org>
+PKG_LICENSE:=GPL-2.0
```
Comment (BKPepe) [inline net/dawn/Makefile]: This is using deprecated [SPDX License Identifier](https://spdx.org/licenses/).

File: net/dawn/Makefile [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+PKG_LICENSE_FILES:=COPYING
```
Comment (BKPepe) [inline net/dawn/Makefile]: This is wrong. In the repository, there is a file [LICENSE](https://github.com/berlin-open-wireless-lab/DAWN/blob/master/LICENSE).

File: net/dawn/Makefile [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+	SECTION:=net
+	CATEGORY:=Network
+	TITLE:=Decentralized wifi controller.
+	MAINTAINER:=Nick Hainke <vincent@systemli.org>
+	URL:=https://github.com/berlin-open-wireless-lab/DAWN.git
+	DEPENDS:=$(DRV_DEPENDS) +libubus +libubox +libblobmsg-json +libuci +libgcrypt +libiwinfo +umdns
+endef
+
+define Package/dawn/description
+	This package implements a decentralized wireless daemon.
```
Comment (BKPepe) [inline net/dawn/Makefile]: Two spaces as indentation, please.

**Comment by BKPepe** [comment]
> If I should maintain the controller here, I would like commit priviliges. ;)

NAK for now. This is too ambitious.
What's wrong with submitting a pull request? At least we know it was compile tested.  You sent 6 PRs to this repository at all. IMHO, you can maintainer it without commit privileges. 

Was it run tested?

**Comment by PolynomialDivision** [comment]
> Was it run tested?

It was tested by me. And it is not for production use, just for testing. I would call this stadium alpha.
People wanted to test the daemon and give me feedback.

> Why uci config, init script and mac_list are not tracked in this repo, but in yours?

I thought, that this is the way you do things...^^ I really do not care. I have to findout how to manage the config files if they are in this git. I will change this.

> What's wrong with submitting a pull request?

It is just more work for me, to create an extra branch and then do a PR. ;)

**Comment by BKPepe** [comment]
ping @diizzyy

**Comment by PolynomialDivision** [comment]
> ping @diizzyy about the testing stage.

What testing stage?

**Comment by PolynomialDivision** [comment]
@BKPepe, sry I just deleted all my changes cause I forced pushed. I have to switch always between my build server and my local pc... ^^ Now I pushed everything.

**Comment by PolynomialDivision** [comment]
@BKPepe You can not use You can not use 2 spaces for indentation. Otherweise the Makefile will fail to compile?

**Comment by diizzyy** [comment]
Not sure what I'm supposed to do here but what's goal in the end? The only(?) documentation available tells me how to get connected clients listed, list APs and connected clients in "ubus"-format and I presume more functionality supported? Does it really need to start at 50?

File: net/dawn/Makefile [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+PKG_SOURCE_URL:=https://github.com/berlin-open-wireless-lab/DAWN.git
+PKG_SOURCE_VERSION:=f03f55ff92153337972cb9b2d1856b9a364df073
+PKG_MIRROR_HASH:=aaa21efafac48749c0cf233bc723bb42a1d1591761d33cc1cee5b27bed8adee1
+
+PKG_MAINTAINER:=Nick Hainke <vincent@systemli.org>
+PKG_LICENSE:=GPL-2.0
+PKG_LICENSE_FILES:=LICENSE
+
+PKG_BUILD_PARALLEL:=1
+CMAKE_INSTALL:=1
```
Comment (neheb) [inline net/dawn/Makefile]: This is actually useless. cmake.mk defines PKG_INSTALL:=1. CMAKE_INSTALL is PKG_INSTALL + InstallDev, which is not needed here.

File: net/dawn/Makefile (hunk 4 above)
Comment (neheb) [reply net/dawn/Makefile]: More specificallty, this should be -or-later or -only.

File: net/dawn/Makefile [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  DEPENDS:=$(DRV_DEPENDS) +libubus +libubox +libblobmsg-json +libuci +libgcrypt +libiwinfo +umdns
+endef
+
+define Package/dawn/description
+  This package implements a decentralized wireless daemon.
+endef
+
+define Package/dawn/install
+	$(INSTALL_DIR) $(1)/usr/sbin
+	$(INSTALL_BIN) $(PKG_BUILD_DIR)/src/dawn $(1)/usr/sbin/dawn
```
Comment (neheb) [inline net/dawn/Makefile]: Since PKG_INSTALL is 1, this can be switched to PKG_INSTALL_DIR.

File: net/dawn/Makefile (hunk 4 above)
Comment (PolynomialDivision) [reply net/dawn/Makefile]: I have no idea, what a good license is. Any recommendations?

**Comment by PolynomialDivision** [comment]
I added my own opkg feed: https://github.com/berlin-open-wireless-lab/DAWN/issues/53#issuecomment-603487971
Maybe this is a better solution?

File: net/dawn/Makefile [hunk 9]
```diff
@@ -0,0 +1,48 @@
+#
+# This software is licensed under the Public Domain.
+#
+
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=dawn
+PKG_VERSION:=2020-03-23
```
Comment (neheb) [inline net/dawn/Makefile]: Replace with PKG_SOURCE_DATE

File: net/dawn/files/dawn.init [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+	echo "Starting Service..."
+	procd_open_instance
+	procd_set_param command $PROG
+	procd_set_param stdout 0 # here it is possible to remove the debug output...
+	procd_set_param stderr 1
+	procd_add_mdns "dawn" "tcp" "${_tcp_buffer}"
+	procd_close_instance
+  	echo "Dawn instance started!"
+  	echo "UMDNS with port ${_tcp_buffer}"
+}
```
Comment (neheb) [inline net/dawn/files/dawn.init]: No idea if this means it's missing a newline.

Comment (PolynomialDivision) [reply net/dawn/files/dawn.init]: hmmm, I added a newline... ^^

File: net/dawn/Makefile (hunk 9 above)
Comment (PolynomialDivision) [reply net/dawn/Makefile]: Hmmm, I did the same in libuwifi? I have to change it their too? Or will that break everything?

Comment (neheb) [reply net/dawn/Makefile]: No. Just cosmetic.

File: net/dawn/Makefile [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+endef
+
+define Package/dawn/install
+	$(INSTALL_DIR) $(1)/usr/sbin
+	$(INSTALL_BIN) $(PKG_INSTALL_DIR)/usr/sbin/dawn $(1)/usr/sbin/dawn
+	$(INSTALL_DIR) $(1)/etc/init.d
+	$(INSTALL_BIN) ./files/dawn.init $(1)/etc/init.d/dawn
+	$(INSTALL_DIR) $(1)/etc/config
+	$(INSTALL_CONF) ./files/dawn.config $(1)/etc/config/dawn
+	$(INSTALL_DIR) $(1)/tmp
```
Comment (neheb) [inline net/dawn/Makefile]: Uhhh what?

This will get wiped on boot.

If you need an empty file in /tmp, use touch in the init script.

Comment (PolynomialDivision) [reply net/dawn/Makefile]: That is true! Someone demanded that the file should be located in `/tmp/`. That would be a huge problem. Thanks a lot! :)

**Pull request merged**